metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py

```diff
@@ -0,0 +1,299 @@
+        scaling = head_dim ** -0.5
+        grad_mha_output= grad_mha_output.view(seq_len, -1, head_dim).transpose(0, 1)
+        grad_v = torch.bmm(attn_probs.transpose(1,2), grad_mha_output).transpose(0, 1).contiguous().view(seq_len, bsz, -1)
+        grad_attn_probs_out = torch.bmm(grad_mha_output, v.transpose(1, 2))
+
+        grad_attn_probs_in = scaled_upper_triang_masked_softmax_cuda.backward(
+            grad_attn_probs_out, attn_probs, 1.0
+        )
+        grad_q = torch.bmm(
+            math.sqrt(scaling) * grad_attn_probs_in,
+            math.sqrt(scaling) * k.transpose(0,1)
+        )
+        grad_q = grad_q.transpose(0, 1).contiguous().view(seq_len, bsz, -1)
+        grad_k = torch.bmm(
+            math.sqrt(scaling) * grad_attn_probs_in.transpose(1,2),
+            math.sqrt(scaling) * q.transpose(0,1)
+        )
+        grad_k = grad_k.transpose(0, 1).contiguous().view(seq_len, bsz, -1)
+        grad_kvq_proj_output = torch.cat([grad_k, grad_v, grad_q], dim=-1)
+        return grad_kvq_proj_output
+
+    @staticmethod
+    def forward(
+        ctx,
+        input,
+        kvq_proj_weight,
+        out_proj_weight,
+        fc1_weight,
+        fc2_weight,
+        head_dim,
+        recompute_fc1,
+    ):
+        ctx.recompute_fc1 = recompute_fc1
+
+        input = input.contiguous()
+
+        # Take out residual connection for self attention
+        residual = input
+        dtype = input.dtype
+
+        # Apply layer norm on (seq_len // #tp_size, bsz, embed_dim) tensor
+        ctx.layer_norm_normalized_shape = torch.Size((input.size(-1),))
+        ctx.eps = 1e-5
+
+        # # Self attention layer norm
+        mha_layer_norm_output, _, _ = fused_layer_norm_cuda.forward(input, ctx.layer_norm_normalized_shape, ctx.eps)
+
+        # all gather output across first dim, i.e. seq_len dim for kvq_proj
+        mha_layer_norm_output = _gather_along_first_dim(mha_layer_norm_output, cached_buffer_name='mpu')
+
+        # apply kvq, output is (seq_len, bsz, 3 * embed_dim // #tp_size)
+        kvq_out = torch.matmul(mha_layer_norm_output, kvq_proj_weight.t())
+        # the order here doesn't matter as much as long its consistent sice initialization is same.
+        # just matching the ordewr of metaseq MHA.
+        k, v, q = split_tensor_along_last_dim(kvq_out, 3, contiguous_split_chunks=True)
+        seq_len, bsz, embed_dim_per_partition = q.size()
+        q  = q.view(seq_len, -1, head_dim)
+        k  = k.view(seq_len, -1, head_dim)
+        v  = v.view(seq_len, -1, head_dim).transpose(0, 1)
+
+        attn, _ = SequeuceParallelTransformerBlock.forward_mha(q, k, v, bsz, seq_len, head_dim, embed_dim_per_partition, dtype)
+
+        out_proj_out = torch.matmul(attn, out_proj_weight.t())
+        out_proj_out = _reduce_scatter_along_first_dim(out_proj_out)
+
+        # out_proj_out = out_proj_out + residual
+        out_proj_out = out_proj_out + residual
+
+        # Take out residual connection for FFN
```

Comment:
Nit: `out_proj_out` seems to contain residual at this point, which gets passed into FFN... how residual get "taken out" here?

Reply:
Oh maybe `out_proj_out = out_proj_out + residual` is supposed to be removed in Line 128?

Reply:
Oh nvm, taken out == saved